extract.R

```diff
@@ -0,0 +1,173 @@
+semtext <- function(fit, x2=T, cfi=F, tli=T, rmsea=T, rmsea.ci=T, srmr=T, bic=F, aic=F,
+                    clipboard=T, type = "scaled", srmr.type = "srmr", output="text",
+                    modelnames=NULL,
+                    sep = "\t", dec=",") {
+  ## fitted model or list of fitted model
+  ## type: standard , scaled , robust
+  ## srmr.type: srmr, srmr_bentler, srmr_bentler_nomean, srmr_bollen, srmr_bollen_nomean, srmr_mplus, srmr_mplus_nomean
+  
+  if(!is.null(modelnames)) {
+    model.names <- modelnames
+  } else {
+    model.names <- as.list(sys.call())[[2]]
+    model.names <- gsub("list", "", model.names)[-1]
+  }
+  
+  models <- length(fit)
+  fitvalue <- list()
+  
+  ## extract fitmeasuers
+  if (models == 1) {
+    fitvalue[[1]] <- fitmeasures(fit)
+  } else {
+    for (i in c(1:length(fit))) {
+      fitvalue[[i]] <- fitmeasures(fit[[i]])
+    }
+  }
+
+  
+  ## type of robust chi2
+  if (type == "robust") {
+    if(x2 == F) {x2 <- NULL} else {x2 <- c("chisq.scaled", "df.scaled", "pvalue.scaled")}
+    if(cfi == F) {cfi <- NULL} else {cfi <- "cfi.robust"}
+    if(tli == F) {tli <- NULL} else {tli <- "tli.robust"}
+    if(rmsea == F) {rmsea <- NULL} else {rmsea <- "rmsea.robust"}
+    if(rmsea.ci == F) {rmsea.ci <- NULL} else {rmsea.ci <- c("rmsea.ci.lower.robust", "rmsea.ci.upper.robust")}
+    if(srmr == F) {
+      srmr <- NULL
+    } else if (is.null(srmr.type)) {
+      srmr <- srmr
+    } else {srmr <- srmr.type}
+    if(bic == F) {bic <- NULL} else {bic <- "bic"}
+    if(aic == F) {aic <- NULL} else {aic <- "aic"}
+  }
+  else if (type == "scaled") {
+    if(x2 == F) {x2 <- NULL} else {x2 <- c("chisq.scaled", "df.scaled", "pvalue.scaled")}
+    if(cfi == F) {cfi <- NULL} else {cfi <- "cfi.scaled"}
+    if(tli == F) {tli <- NULL} else {tli <- "tli.scaled"}
+    if(rmsea == F) {rmsea <- NULL} else {rmsea <- "rmsea.scaled"}
+    if(rmsea.ci == F) {rmsea.ci <- NULL} else {rmsea.ci <- c("rmsea.ci.lower.scaled", "rmsea.ci.upper.scaled")}
+    if(srmr == F) {
+      srmr <- NULL
+    } else if (is.null(srmr.type)) {
+      srmr <- srmr
+    } else {srmr <- srmr.type}
+    if(bic == F) {bic <- NULL} else {bic <- "bic"}
+    if(aic == F) {aic <- NULL} else {aic <- "aic"}
+  } else {
+    if(x2 == F) {x2 <- NULL} else {x2 <- c("chisq", "df", "pvalue")}
+    if(cfi == F) {cfi <- NULL} else {cfi <- "cfi"}
+    if(tli == F) {tli <- NULL} else {tli <- "tli"}
+    if(rmsea == F) {rmsea <- NULL} else {rmsea <- "rmsea"}
+    if(rmsea.ci == F) {rmsea.ci <- NULL} else {rmsea.ci <- c("rmsea.ci.lower", "rmsea.ci.upper")}
+    if(srmr == F) {
+      srmr <- NULL
+    } else if (is.null(srmr.type)) {
+      srmr <- srmr
+    } else {srmr <- srmr.type}
+    if(bic == F) {bic <- NULL} else {bic <- "bic"}
+    if(aic == F) {aic <- NULL} else {aic <- "aic"}
```

Comment:
Hele, tohleto mi vsechno prijde strasne tezkopadne, hrozne blbe se to cte a pokud bys to potrebovat nejak predelavat, tak se z toho nejspis poseres. Napadl Te nejaky lepsi zpusob, jak to delat? Mne napada tohleto:

- Ted vypada definice funkce nasledovne:
```function(fit, x2=T, cfi=F, tli=T, rmsea=T, rmsea.ci=T, srmr=T, bic=F, aic=F,
               clipboard=T, type = "scaled", srmr.type = "srmr", ...){}```

...user musi zapinat indexy jeden po druhem (coz mi prijde dost neprakticky) a Ty pak musis uvnitr funkce solichat s kazdym argumentem. Co takhle:

```function(fit, indices = c('x2', 'cfi', 'tli', ...), 
                    clipboard=T, type = "scaled",  ...){}```

...user by pouzil bud defaultni vektor `indices` (obsahoval by to stejne co mas ted jako TRUE) a nebo by si napsal vlastni. Pak by to fungovalo takto:

1) podle vstupu do `type` si napises vektor nazvu indexu, jako kdyby user chtel vsechny
2) pomoci toho vektoru ^ extrahujes z `fitmeasures()` vse, co by mohl user pro dany `type` potrebovat a vlozis do `df`
3) prejmenujes sloupce `df` tak, aby korespondovaly s tim, co bude validni input do `indices`
4) pomoci `intersect()` vyfiltrujes z `df` jen ty sloupce, o ktere si rekl user v `indices`
5) Finalni text pak naparsujes tak, ze budes skladat nazev sloupce, '=', a hodnotu ve sloupci

dava smysl?

Reply:
Jo, to je mnohem lepší! Nějak takhle jsem nad tím přemýšlel a chtěl jsem to v tom duchu předělávat. Blbé je, že to nebude kompatibilní s předchozími verzemi té funkce :( Takže to tam bude chtít dát nějaký warning, protože to používám v některých skriptech (které jsem třeba  i sdílel), co myslíš? A stejně tak ten `output = "table"` nedává moc smysl, lepší je `table=T`, co?

Reply:
V pohode, ja ten kod upravim, jen to nechci delat bez svoleni a vyjasneni :) Jen mi potvrd, ze jsi na tom nezacal delat.

Reply:
O kompatibilitu prijdes, no, jak moc velky je to problem? Muzeme tu funkci nechat ve stavajici podobe a vytvorit novou, s nejakym privlastkem, cimz minimalne pro nekolik dalsich verzi budes davat najevo, ze s tou starou funkci uz nepocitas. Do dokumentaci taky napiseme, ze ta funkce pujde pomalu pryc.